crates/utils/re_auth/src/oauth/api.rs

```diff
@@ -19,88 +24,275 @@ pub enum Error {
-    endpoint: impl std::fmt::Display,
-) -> Result<Res, Error> {
-    send(ehttp::Request::get(format!(
-        "{base_url}{endpoint}",
-        base_url = *API_BASE_URL
-    )))
-    .await
-}
+pub trait IntoRequest: Sized {
+    type Res: serde::de::DeserializeOwned;
 
-async fn post<Body: serde::Serialize, Res: serde::de::DeserializeOwned>(
-    endpoint: impl std::fmt::Display,
-    body: Body,
-) -> Result<Res, Error> {
-    send(
-        ehttp::Request::json(
-            format!("{base_url}{endpoint}", base_url = *API_BASE_URL),
-            &body,
-        )
-        .map_err(Error::Deserialize)?,
-    )
-    .await
+    fn into_request(self) -> Result<ehttp::Request, Error>;
 }
 
+// NOTE: We use PKCE, so refresh token doesn't require client secret.
 #[derive(serde::Serialize)]
-#[serde(rename_all = "camelCase")]
-struct RefreshRequest {
-    refresh_token: String,
+pub struct AuthenticateWithRefresh<'a> {
+    grant_type: &'a str,
+    client_id: &'a str,
+    refresh_token: &'a str,
+}
+
+impl IntoRequest for AuthenticateWithRefresh<'_> {
+    type Res = RefreshResponse;
+
+    fn into_request(self) -> Result<ehttp::Request, Error> {
+        ehttp::Request::json(
+            format_args!("{base}/user_management/authenticate", base = *WORKOS_API),
+            &self,
+        )
+        .map_err(Error::Serialize)
+    }
 }
 
 #[derive(serde::Deserialize)]
-#[serde(rename_all = "camelCase")]
-pub struct AuthenticationResponse {
-    pub user: User,
+pub struct RefreshResponse {
+    pub user: super::User,
     pub organization_id: Option<String>,
     pub access_token: String,
     pub refresh_token: String,
 }
 
-pub(crate) async fn refresh(refresh_token: &RefreshToken) -> Result<AuthenticationResponse, Error> {
-    post(
-        "/refresh",
-        RefreshRequest {
-            refresh_token: refresh_token.0.clone(),
-        },
-    )
+pub(crate) async fn refresh(refresh_token: &RefreshToken) -> Result<RefreshResponse, Error> {
+    send_async(AuthenticateWithRefresh {
+        grant_type: "refresh_token",
+        client_id: &OAUTH_CLIENT_ID,
+        refresh_token: &refresh_token.0,
+    })
     .await
 }
 
+struct JwksRequest;
+
+impl IntoRequest for JwksRequest {
+    type Res = jsonwebtoken::jwk::JwkSet;
+
+    fn into_request(self) -> Result<ehttp::Request, Error> {
+        Ok(ehttp::Request::get(format_args!(
+            "{base}/sso/jwks/{client_id}",
+            base = *WORKOS_API,
+            client_id = *OAUTH_CLIENT_ID,
+        )))
+    }
+}
+
 pub async fn jwks() -> Result<jsonwebtoken::jwk::JwkSet, Error> {
-    get("/jwks").await
+    send_async(JwksRequest).await
+}
+
+pub struct Pkce {
+    /// random string of bytes
+    code_verifier: String,
+
+    /// base64-encoded sha256 hash of `code_verifier`
+    code_challenge: String,
+}
+
+const CHARSET: &[u8] = b"\
+    ABCDEFGHIJKLMNOPQRSTUVWXYZ\
+    abcdefghijklmnopqrstuvwxyz\
+    0123456789-.~_\
+";
+
+impl Pkce {
```

Comment:
Nice that you implemented it yourself instead of adding an entire crate :)

Reply:
I found a crate that implements this in roughly as many lines in much the same way, which gave me confidence to just implement it myself :sweat_smile: